File: 3952-maximum-total-value-of-covered-indices/3952-maximum-total-value-of-covered-indices.php
class Solution {

    /**
     * @param Integer[] $nums
     * @param String $s
     * @return Integer
     */
    function maxTotal($nums, $s) {
        $neg = -1000000000000000000;

        $dp0 = 0;
        $dp1 = $neg;

        $n = count($nums);

        for ($i = 0; $i < $n; $i++) {
            $nxt0 = $neg;
            $nxt1 = $neg;

            if ($s[$i] === '0') {
                $nxt0 = max($dp0, $dp1);
            } else {
                $nxt1 = max($dp0, $dp1) + $nums[$i];

                if ($i > 0) {
                    $nxt0 = $dp0 + $nums[$i - 1];
                }
            }

            $dp0 = $nxt0;
            $dp1 = $nxt1;
        }

        return max($dp0, $dp1);
    }
}

File: 3952-maximum-total-value-of-covered-indices/3952-maximum-total-value-of-covered-indices.rb
# @param {Integer[]} nums
# @param {String} s
# @return {Integer}
def max_total(nums, s)
  neg = -10**18

  dp0 = 0
  dp1 = neg

  (0...nums.length).each do |i|
    nxt0 = neg
    nxt1 = neg

    if s[i] == '0'
      nxt0 = [dp0, dp1].max
    else
      nxt1 = [dp0, dp1].max + nums[i]

      if i > 0
        nxt0 = dp0 + nums[i - 1]
      end
    end

    dp0 = nxt0
    dp1 = nxt1
  end

  [dp0, dp1].max
end